autogen/oai/cohere.py

```diff
@@ -0,0 +1,447 @@
+        cohere_params["p"] = validate_parameter(params, "p", (int, float), False, 0.75, (0.01, 0.99), None)
+        cohere_params["seed"] = validate_parameter(params, "seed", int, True, None, None, None)
+        cohere_params["frequency_penalty"] = validate_parameter(
+            params, "frequency_penalty", (int, float), True, 0, (0, 1), None
+        )
+        cohere_params["presence_penalty"] = validate_parameter(
+            params, "presence_penalty", (int, float), True, 0, (0, 1), None
+        )
+
+        # Cohere parameters we are ignoring:
+        # preamble - we will put the system prompt in here.
+        # parallel_tool_calls (defaults to True), perfect as is.
+        # conversation_id - allows resuming a previous conversation, we don't support this.
+        logging.info("Conversation ID: %s", params.get("conversation_id", "None"))
+        # connectors - allows web search or other custom connectors, not implementing for now but could be useful in the future.
+        logging.info("Connectors: %s", params.get("connectors", "None"))
+        # search_queries_only - to control whether only search queries are used, we're not using connectors so ignoring.
+        # documents - a list of documents that can be used to support the chat. Perhaps useful in the future for RAG.
+        # citation_quality - used for RAG flows and dependent on other parameters we're ignoring.
+        # max_input_tokens - limits input tokens, not needed.
+        logging.info("Max Input Tokens: %s", params.get("max_input_tokens", "None"))
+        # stop_sequences - used to stop generation, not needed.
+        logging.info("Stop Sequences: %s", params.get("stop_sequences", "None"))
+
+        return cohere_params
+
+    def create(self, params: Dict) -> ChatCompletion:
+
+        messages = params.get("messages", [])
+
+        # Parse parameters to the Cohere API's parameters
+        cohere_params = self.parse_params(params)
+
+        # Convert AutoGen messages to Cohere messages
+        cohere_messages, preamble, final_message = oai_messages_to_cohere_messages(messages, params, cohere_params)
+
+        cohere_params["chat_history"] = cohere_messages
+        cohere_params["message"] = final_message
+        cohere_params["preamble"] = preamble
+
+        # We use chat model by default
+        client = Cohere(api_key=self.api_key)
+
+        # Token counts will be returned
+        prompt_tokens = 0
+        completion_tokens = 0
+        total_tokens = 0
+
+        # Stream if in parameters
+        streaming = True if "stream" in params and params["stream"] else False
+        cohere_finish = ""
+
+        max_retries = 5
+        for attempt in range(max_retries):
+            ans = None
+            try:
+                if streaming:
+                    response = client.chat_stream(**cohere_params)
+                else:
+                    response = client.chat(**cohere_params)
+            except Exception as e:
```

Comment:
Can we use a more specific exception class? We typically avoid using the generic Exception class.

Reply:
updated with a Custom Exception `CohereRateLimitError`. thanks for reviewing @sonichi